docs/FlowClassifyHook.md

```diff
@@ -0,0 +1,302 @@
+        - **Allow**: The connection is deemed safe and permitted to proceed without further inspection.
+        - **Block**: Terminates the connection immediately.
+        - **Need More Data**: Indicates the need for additional data for a conclusive decision. The extension allows the current segment and will invoke the program again for the next segment.
+5. **Stream Inspection**
+    - The eBPF program performs parsing/inspection tasks on the stream.
+    - If the “Need More Data” return code is used, the inspection until a definitive decision is returned.
+6. **Connection Termination or Continuation**
+    - Once a decision other than "Need More Data" is reached (Allow or Block), the lifecycle for that connection ends within the context of the filter program.
+    - For “Block,” the connection is terminated.
+    - For “Allow,” the connection proceeds without further filtering.
+    - If a connection is terminated while still being classified, the `flow_deleted` hook is invoked.
+
+_Notes:_
+
+- The [Windows Stream Inspection](https://learn.microsoft.com/en-us/windows-hardware/drivers/network/stream-inspection)
+API is used for inspecting the stream-layer data via WFP callouts.
+- the WFP `CONDITIONAL_ON_FLOW` flag means the stream-layer callout will only run for flows where the stream-layer WFP context (containing the bpf program context) is set.
+  - The context is freed immediately if the flow is allowed at the flow established layer.
+  - Otherwise the context is freed when an allow/block decision is made or the flow is deleted.
+- When returning NEED_MORE_DATA, the current data segment is allowed (but the hook will keep getting invoked).
+- Multiple eBPF programs can be invoked for a single stream, but the decision is finalized based on the first conclusive return code.
+- No stream mutation is supported through these hooks (only inspection and allow/block).
+
+### WFP Implementation Details
+
+The stream inspection extension integrates with the Windows Filtering Platform (WFP)
+at the flow_established layer to indicate flows for classification and at the stream layer to
+classify flows.
+
+Currently only inline stream classification will be supported.
+
+#### WFP Action Mapping
+
+The hook implementation maps eBPF program return values to specific WFP actions:
+
+- **FLOW_CLASSIFY_ALLOW**: Returns `FWP_ACTION_PERMIT` and deletes the WFP flow context. This allows the current segment and terminates further classification for this flow by removing the flow context.
+
+- **FLOW_CLASSIFY_BLOCK**: Returns `FWP_ACTION_BLOCK` with the `FWPS_CLASSIFY_OUT_FLAG_ABSORB` flag set. This blocks the current segment and terminates the connection by dropping all subsequent packets.
+
+- **FLOW_CLASSIFY_NEED_MORE_DATA**: Returns `FWP_ACTION_PERMIT`. This allows the current segment to proceed while maintaining the flow context for continued classification of subsequent segments.
```

Comment:
Is there a reason why we do this, as opposed to using the FWPS_STREAM_ACTION_NEED_MORE_DATA?